File: sdk/vbnet/CardAuthSDK/Utils/SignHelper.vb
Imports System
Imports System.Collections.Generic
Imports System.Security.Cryptography
Imports System.Text

Namespace CardAuthSDK.Utils
    Public Module SignHelper
        Public Function GenerateNonce() As String
            Return Guid.NewGuid().ToString("N").Substring(0, 16)
        End Function

        Public Function GetTimestamp() As String
            Return (CInt(DateTime.UtcNow.Subtract(New DateTime(1970, 1, 1)).TotalSeconds)).ToString()
        End Function

        Public Function HmacSha256(data As String, key As String) As String
            Using hmac As New HMACSHA256(Encoding.UTF8.GetBytes(key))
                Dim hashBytes As Byte() = hmac.ComputeHash(Encoding.UTF8.GetBytes(data))
                Dim sb As New StringBuilder()
                For i As Integer = 0 To hashBytes.Length - 1
                    sb.Append(hashBytes(i).ToString("x2"))
                Next
                Return sb.ToString()
            End Using
        End Function

        Public Function Sign(method As String, path As String, timestamp As String, nonce As String, body As String, appSecret As String) As String
            Dim signString As String = method.ToUpper() + path + timestamp + nonce + body
            Return HmacSha256(signString, appSecret)
        End Function
    End Module
End Namespace

File: sdk/csharp/CardAuthSDK/Utils/SignHelper.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;

namespace CardAuthSDK.Utils
{
    public static class SignHelper
    {
        public static string GenerateNonce()
        {
            return Guid.NewGuid().ToString("N").Substring(0, 16);
        }

        public static string GetTimestamp()
        {
            return ((int)(DateTime.UtcNow.Subtract(new DateTime(1970, 1, 1))).TotalSeconds).ToString();
        }

        public static string HmacSha256(string data, string key)
        {
            using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(key)))
            {
                byte[] hashBytes = hmac.ComputeHash(Encoding.UTF8.GetBytes(data));
                StringBuilder sb = new StringBuilder();
                for (int i = 0; i < hashBytes.Length; i++)
                {
                    sb.Append(hashBytes[i].ToString("x2"));
                }
                return sb.ToString();
            }
        }

        public static string Sign(string method, string path, string timestamp, string nonce, string body, string appSecret)
        {
            string signString = method.ToUpper() + path + timestamp + nonce + body;
            return HmacSha256(signString, appSecret);
        }
    }
}
